File: assets/downloads/Nonlinear_Control/Phase_plane_plot.m
function phase_plane_saturation()
    % Phase plane analysis of the nonlinear system with saturation
    % dot{x1} = x1 + sat(-9/2*x1 + 1/2*x2)
    % dot{x2} = -x2 + sat(-9/2*x1 + 1/2*x2)

    % Define grid for vector field
    xmin = -4; xmax = 4;
    ymin = -3; ymax = 3;
    nx = 25; ny = 25;
    [X1, X2] = meshgrid(linspace(xmin, xmax, nx), linspace(ymin, ymax, ny));

    % Vector field
    U = X1 + sat(-4.5 * X1 + 0.5 * X2);
    V = -X2 + sat(-4.5 * X1 + 0.5 * X2);

    % Normalize arrows for nicer quiver plot
    M = hypot(U, V);
    M(M == 0) = 1;
    U2 = U ./ M;
    V2 = V ./ M;

    % Plot setup
    figure('Position', [200 100 1000 600]); hold on;
    quiver(X1, X2, U2, V2, 0.6, 'Color', [0.5 0.5 0.5]);
    colormap('parula'); % ✅ Changed from 'plasma' to 'parula'

    xlabel('x_1', 'FontSize', 14);
    ylabel('x_2', 'FontSize', 14);
    title('Phase plane: vector field + trajectories', 'FontSize', 18);
    xlim([xmin xmax]);
    ylim([ymin ymax]);
    grid on;

    % Compute and plot nullclines (dx/dt = 0 and dy/dt = 0)
    DX = U;
    DY = V;
    contour(X1, X2, DX, [0 0], 'r--', 'LineWidth', 2);
    contour(X1, X2, DY, [0 0], 'b--', 'LineWidth', 2);

    % List of initial conditions
    ic_list = [
        -0.5,  2.4;
        -1.01, -1.5;
        -0.99, -1.5;
         0,   -2.75;
         0,   -1.0;
         1.001, 2;
         0.999, 2
    ];

    % Time span for integration
    tspan = [0 40];

    % Simulate trajectories
    for k = 1:size(ic_list,1)
        ic = ic_list(k,:);
        [t, z] = ode45(@(t, z) f(t, z), tspan, ic);
        plot(z(:,1), z(:,2), '-k', 'LineWidth', 1);
        plot(ic(1), ic(2), 'xr', 'LineWidth', 2, 'MarkerSize', 8);
    end

    % Legend
    legend({'Vector field', 'dx1/dt = 0', 'dx2/dt = 0', ...
        'Trajectories', 'Start points'}, ...
        'Location', 'northeastoutside');

    hold off;

    % Save figure
    saveas(gcf, 'Phase_plane_plot_MATLAB.png');
end

%-------------------------------------------------------------
% System dynamics
function dz = f(~, z)
    x1 = z(1);
    x2 = z(2);
    u = -4.5 * x1 + 0.5 * x2;
    dx1 = x1 + sat(u);
    dx2 = -x2 + sat(u);
    dz = [dx1; dx2];
end

%-------------------------------------------------------------
% Saturation function
function y = sat(u)
    y = min(max(u, -1), 1);
end

File: assets/downloads/Nonlinear_Control/Phase_plane_plot.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

def sat(u):
    """Saturation function: clips input to the range [-1, 1]."""
    return np.clip(u, -1.0, 1.0)

def f(t, z):
    """Defines the system of ODEs (signature compatible with solve_ivp)."""
    x1, x2 = z
    u = -9/2 * x1 + 1/2 * x2
    dx1 = x1 + sat(u)
    dx2 = -x2 + sat(u)
    return [dx1, dx2]

# grid for vector field
xmin, xmax, ymin, ymax = -4, 4, -3, 3
nx, ny = 25, 25
x1 = np.linspace(xmin, xmax, nx)
x2 = np.linspace(ymin, ymax, ny)
X1, X2 = np.meshgrid(x1, x2)

# vector field for the specified system:
U = X1 + sat(-4.5 * X1 + 0.5 * X2)
V = -X2 + sat(-4.5 * X1 + 0.5 * X2)

# normalize arrows for nicer display
M = np.hypot(U, V)
M[M == 0] = 1.0
U2, V2 = U / M, V / M

fig, ax = plt.subplots(figsize=(15, 8))
ax.set_xlim(xmin, xmax)
ax.set_ylim(ymin, ymax)
ax.set_xlabel('x1', fontsize=14)
ax.set_ylabel('x2', fontsize=14)
ax.set_title('Phase plane: vector field + trajectories', fontsize=20)

# quiver (sparse arrows)
ax.quiver(X1, X2, U2, V2, M, pivot='mid', cmap='plasma', alpha=0.75)

# nullclines: plot contours where dx/dt = 0 and dy/dt = 0 (computed on grid)
DX = U  # dx/dt on grid
DY = V  # dy/dt on grid
ax.contour(X1, X2, DX, levels=[0], colors='r', linestyles='--', linewidths=2)
ax.contour(X1, X2, DY, levels=[0], colors='b', linestyles='--', linewidths=2)

# sample trajectories from several initial conditions (original list)
ic_list = [[-0.5, 2.4], 
           [-1.01, -1.5], 
           [-0.99, -1.5],
           [0, -2.75], 
           [0, -1.0], 
           [1.001, 2], 
           [0.999, 2]]

# plot trajectories (shorter time for illustration)
tspan = (0, 40)
t_eval = np.linspace(tspan[0], tspan[1], 2000)
for ic in ic_list:
    sol = solve_ivp(f, tspan, ic, t_eval=t_eval, rtol=1e-6)
    ax.plot(sol.y[0], sol.y[1], '-k', lw=1, alpha=0.9)

# mark the starting points of each trajectory with a red cross
for ic in ic_list:
    ax.plot(ic[0], ic[1], marker='x', color='red', markersize=8, mew=2, linestyle='None')

# create legend entries for nullclines, trajectories, and starting points
ax.plot([], [], 'r--', lw=2, label='dx1/dt = 0 (nullcline)')
ax.plot([], [], 'b--', lw=2, label='dx2/dt = 0 (nullcline)')
ax.plot([], [], '-k', label='trajectories')
ax.plot([], [], marker='x', color='red', linestyle='None', label='trajectory start')
ax.legend(loc='upper right', fontsize=14)

ax.grid(alpha=0.4)
plt.tight_layout()
plt.savefig('ch2_ex1_phase_plane.png', dpi=300)
# plt.show()
plt.close()